### ISSUE TITLE
Cursor advances when terminal is resized (sometimes)

### ISSUE THREAD

**Issue by mattfbacon** [fish-shell/fish-shell#8247]
fish 3.3.1    
Arch Linux    
kitty terminal 0.23.1    

Unfortunately I can't reproduce this consistently so no Asciinema for now. I just had it happen a few minutes ago, and it's been happening fairly often in the past while. Basically what happens is that the cursor advances when the terminal window is resized. When I press any key, the cursor moves back to the correct position.

**Comment by mattfbacon** [comment]
I have reproduced it.

![Demo of issue](https://i.imgur.com/U5ONcYS.gif)

**Comment by faho** [comment]
Fish [disables its reflow handling in kitty](https://github.com/fish-shell/fish-shell/blob/48e4ce2f6d05e207cbfbacc5c13a05612fe702c9/share/functions/__fish_config_interactive.fish#L239), relying on kitty to handle it (otherwise we'd have weird issues related to both racing against each other). It still repaints once it is all done. It seems in this case kitty moves the cursor, there is nothing fish can do.

**Comment by mattfbacon** [comment]
Hmm, let me see if I can reproduce it with another shell.

**Comment by kovidgoyal** [comment]
@mattfbacon See https://github.com/kovidgoyal/kitty/issues/3948  with that kitty no longer reflows the active prompt, instead it clears it and expects the shell to redraw it. 

@faho: If you want to avoid races when redrawing the prompt, use the escape sequence for atomic updates to draw the prompt: https://github.com/kovidgoyal/kitty/issues/2145#issuecomment-556726764
and assume that a reflow always leaves the vertical position of the cursor with respect to the first prompt line unchanged. So all you have to do is move the cursor up that number fo lines, and then to the zeroth column and clear to end of screen and redraw the prompt.

**Comment by faho** [comment]
@kovidgoyal Can't say I love the special escape sequence - those are awkward to detect support for. 

How exactly are we supposed to do that in this case? Just "$TERM = kitty" or "iTERM"? That's how these things always fall flat.

**Comment by mattfbacon** [comment]
So, to be clear, fish expects kitty to reflow, and kitty expects fish to reflow? LOL

**Comment by kovidgoyal** [comment]
You dont need to detect support for it. Emitting it is harmless it will simply have no effect on terminals that dont support it. In general, *never* use TERM to detect things, its beyond fragile.

@mattfbacon kitty's shell integration turns on fishs prompt redraw behavior automatically.

**Comment by faho** [comment]
>So, to be clear, fish expects kitty to reflow, and kitty expects fish to reflow? LOL

Used to be kitty reflowed, so we turned it off, now kitty apparently doesn't reflow any longer.

@mattfbacon `set -g fish_handle_reflow 1`.

>You dont need to detect support for it. Emitting it is harmless it will simply have no effect on terminals that dont support it.

That's... optimistic.

> In general, never use TERM to detect things, its beyond fragile.

I am aware. I've done a lot of terminal detection. It sucks. And somehow users insist on using bad terminals.

**Comment by mattfbacon** [comment]
> That's... optimistic.

Agreed. I recently saw an issue about a terminal randomly adding h, l, and some other character. Turned out it was escape sequences.

> @mattfbacon `set -g fish_handle_reflow 1`.

Right, that's what I was thinking. I will reply back if there are any issues after setting that.

**Comment by kovidgoyal** [comment]
> > So, to be clear, fish expects kitty to reflow, and kitty expects fish to reflow? LOL
> 
> Used to be kitty reflowed, so we turned it off, now kitty apparently doesn't reflow any longer.
> 
> @mattfbacon `set -g fish_handle_reflow 1`.

No this is in master not the version of kitty @mattfbacon is running.

> 
> > You dont need to detect support for it. Emitting it is harmless it will simply have no effect on terminals that dont support it.
> 
> That's... optimistic.

No, it isnt. It was extensively tested. In the original DCS form only the linux kernel console and apple's terminal had issues with it and in the later CSI form, no terminals have issues with it.

**Comment by faho** [comment]
>No this is in master not the version of kitty @mattfbacon is running.

Then what is @mattfbacon seeing now?

> No, it isnt. It was extensively tested. In the original DCS form only the linux kernel console and apple's terminal had issues with it and in the later CSI form, no terminals have issues with it. But, suit yourself.

That's good to know, then. I assumed you were referring to kitty, iTerm and some nebulous "standard" that doesn't exist in practice.

Is this variable anywhere? Any idea of adding this to terminfo so other terminals can pick a different sequence (for some reason)? Or is it just that actual thing, always, that we can just send?

**Comment by kovidgoyal** [comment]
> > No this is in master not the version of kitty @mattfbacon is running.
> 
> Then what is @mattfbacon seeing now?

No idea, whatever it is he will stop seeing it with shell integration so am not motivated to investigate.

> 
> > No, it isnt. It was extensively tested. In the original DCS form only the linux kernel console and apple's terminal had issues with it and in the later CSI form, no terminals have issues with it. But, suit yourself.
> 
> That's good to know, then. I assumed you were referring to kitty, iTerm and some nebulous "standard" that doesn't exist in practice.
> 
> Is this variable anywhere? Any idea of adding this to terminfo so other terminals can pick a different sequence (for some reason)? Or is it just that actual thing, always, that we can just send?

There are two forms of it only, the original one based on DCS and a newer one based on CSI. Either works, just pick one. I suggest the CSI form to avoid issues with apple's terminal. Emit CSI2026h before drawing anything and CSI2026l after you are done drawing.

**Comment by mattfbacon** [comment]
> No idea, whatever it is he will stop seeing it with shell integration so am not motivated to investigate.

@kovidgoyal What is the timeline on this?

**Comment by kovidgoyal** [comment]
> > No idea, whatever it is he will stop seeing it with shell integration so am not motivated to investigate.
> 
> @kovidgoyal What is the timeline on this?

There is no timeline. its already implemented. Just run kitty from master, or subscribe to the issue I linked to earlier to be notified when it is released.

**Comment by mattfbacon** [comment]
I just meant that it's not in a release currently, correct? Arch Linux packages generally follow releases so I was more curious when the next release (containing that feature) will be released.

**Comment by kovidgoyal** [comment]
Yes its not released. As for when the next kitty release will be, probably a month or two, though I dont do any form of release scheduling, its just when I feel master is in a "good" state.

**Comment by mattfbacon** [comment]
Alright, in that case, since this issue is (theoretically) fixed in master, we can close this issue?

Thanks @faho and @kovidgoyal. I love fish and kitty and appreciate your responsiveness.

**Comment by faho** [comment]
Okay, I do not believe adding this works unless we have some way to detect support.

>In the original DCS form only the linux kernel console and apple's terminal had issues with it and in the later CSI form, no terminals have issues with it.

The idea that all terminals would just ignore unknown DECSET style (`\e[?`, then some number and then "h" or "l") sequences is incorrect - that's how we get Windows' conpty printing "l" and "h" junk for bracketed paste (#8240 and duplicates). It's also not true that it'll be quickly fixed, this has been an issue for months now.

Sorry.

**Comment by kovidgoyal** [comment]
I dont follow. You already use DECSET for bracketed paste. Simply detect if you are running on windows and disable till Microsoft get their act together. Is it really your position that you will never use DECSET ever again? Because of a bug in conpty? That is going to seriously crimp the future development of fish.

**Comment by faho** [comment]
No, what I'm saying is that we don't want to use DECSET *without checking*.

> Simply detect if you are running on windows  

This is WSL. Fish isn't running "on windows", and the terminal can even be something reasonable like alacritty. 

We need to check for WSL specifically, which is possible. 

I'm simply saying: Your tests clearly weren't exhaustive, so I want to be more careful.

**Comment by kovidgoyal** [comment]
On Sat, Feb 19, 2022 at 07:19:36AM -0800, Fabian Homborg wrote:
> No, what I'm saying is that we don't want to use DECSET *without checking*.

Send a DECQRM followed immediately by a carriage return and
and erase to end of line. If you get a response to decqrm the terminal
supports decset. If it doesnt the carriage return and erase to end of
line will remove any junk that results.

This has the downside that you cannot use DECSET based features until
after you receive the response.

Though if I were you I would simply disable on windows and run it
blindly everywhere else.

> > Simply detect if you are running on windows  
> 
> This is WSL. Fish isn't running "on windows", and the terminal can even be something reasonable like alacritty. 

WSL is still windows underneath, and I am sure can be detected. 

> 
> We need to check for WSL specifically, which is possible. 

Indeed, I would be amazed if that were not possible.

> 
> I'm simply saying: Your tests clearly weren't exhaustive, so I want to be more careful. 

Most definitely, I have not tested every single terminal ever created. I
wont even say I have tested most of them since that would require a time
machine. But you dont need to rely on my tests since you already use
DECSET for bracketed paste and your own userbase has told you that only
conpty doesnt support it.

**Comment by faho** [comment]
> Send a DECQRM followed immediately by a carriage return
 
I have explained this to you before, querying doesn't work for us.

It's slow, doesn't scale and *still* requires the terminal to not mess up.

I can see how it seems appealing from a terminal authors perspective, but from our POV it's simply unusable.

Users use horrible terminals, and when something goes wrong they blame us - often with a comment that it "worked in bash". Of course that is because bash barely exercises the terminal - it would, for the most part, be usable on a printer.

**Comment by mattfbacon** [comment]
Is it unreasonable to expect terminals to swallow escape sequences they don't understand, rather than printing junk?

**Comment by kovidgoyal** [comment]
On Sat, Feb 19, 2022 at 04:43:58PM -0800, Matt Fellenz wrote:
> Is it unreasonable to expect terminals to swallow escape sequences they don't understand, rather than printing junk?

Nope its not. Implementing a basic VT parser is table stakes for a
terminal. And I am firmly of the opinion that terminal programs should
rely on that.

**Comment by kovidgoyal** [comment]
On Sat, Feb 19, 2022 at 02:29:43PM -0800, Fabian Homborg wrote:
> > Send a DECQRM followed immediately by a carriage return
>  
> I have explained this to you before, querying doesn't work for us.
> 
> It's slow, doesn't scale and *still* requires the terminal to not mess up.

Except that in this particular case since the functionality you are
querying can simply be turned on *after* you receive a response without
any major effect (the first prompt you draw will be drawn in a non-atomic
manner, that's all). Thus, it's not slow and thanks to the technique I
described doesn't require the terminal to not mess up. Even if the
terminal prints the escape code to STDOUT, it is instantly erased.

I'm not sure what *scaling* refers to here. 

And again, in this case I would simply turn it off on windows/wsl for a
year or two and use it unconditionally everywhere else.

> 
> I can see how it seems appealing from a terminal authors perspective, but from our POV it's simply unusable.
> 
> Users use horrible terminals, and when something goes wrong they blame us - often with a comment that it "worked in bash". Of course that is because bash barely exercises the terminal - it would, for the most part, be usable on a printer. 

Haha yes, bash (more accurately readline) is truly primitive.

But, so what? Users use horrible programs and blame terminal authors all
the time saying something works in X terminal but not in yours. For
example, the new keyboard protocol I developed, causes bug
reports, because vim uses an extremely crazy algorithm to turn on
extended key support.

That's part of the game. If I was afraid of such bug reports, I
would never implement anything new in kitty. In an ecosystem like this
one where the interacting programs are extremely loosely coupled and
exist in the wild in a huge proliferation of implementations and
versions, anything you do is going to cause some breakage somewhere.

Anyway, it's your program, I do not want to waste your time, so all
the best.

**Comment by gnachman** [comment]
iTerm2 version 3.5.0 implements a new mechanism for discovering capabilities which does not rely on reporting (which @faho mentioned is unworkable) that would allow you to add support for synchronization.

The full spec is here: https://iterm2.com/feature-reporting/

tl;dr: look for `Sy` in `$TERM_FEATURES` prior to any non-alphanumeric character. If this is present, then the terminal supports synchronization.
